Add support for DVSPortal (dutch parking) to HA

ChessSpider opened the pull request:
## Proposed change

Add support of DVSPortal. DVSPortal is a parking system in use by some dutch muncipalities. Like these: https://www.google.com/search?q=inurl%3Advsportal 


Documentation PR https://github.com/home-assistant/home-assistant.io/pull/36722

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ChessSpider committed:
dvs portal for ha, initial commit..

ChessSpider committed:
implement more quality checks

ChessSpider committed:
Made another iteration on it.. Next up: cleanup and documentation

ChessSpider committed:
update to typed version of py-dvsportal

ChessSpider committed:
remove unneeded files

ChessSpider commented:
Documentation PR is here https://github.com/home-assistant/home-assistant.io/pull/36722

ChessSpider committed:
also nto needed

joostlek reviewed (changes requested):
Let's split out the service call for now to keep the PR better scoped

ChessSpider committed:
Remove all service calls as requested

ChessSpider commented:
hi @joostlek , im also on discord if you'd like to discuss there. Here in the PR is fine with me  as well. thank you for your time and effort!

Note that im aiming for quality scale bronze now, but i do intent to upgrade to higher quality levels once this is merged in.

ChessSpider commented:
hi @joostlek , just to manage my expectations, when do you thnik youll have time for  a review? thank you in advance

joostlek commented:
I can review when I have the time, there are 480 or so PRs open so there is quite the queue :)

joostlek commented on homeassistant/components/dvsportal/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+        except Exception as e:
+            _LOGGER.exception("Error communicating with API")
+            raise UpdateFailed("Error communicating with API") from e
+        return {
+            "default_code": self.dvs_portal.default_code,
+            "default_type_id": self.dvs_portal.default_type_id,
+            "balance": self.dvs_portal.balance,
+            "active_reservations": self.dvs_portal.active_reservations,
+            "historic_reservations": self.dvs_portal.historic_reservations,
+            "known_license_plates": self.dvs_portal.known_license_plates,
+        }
```
Why don't you use a dataclass? Rather define it here than in types.py

joostlek commented on homeassistant/components/dvsportal/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+            hass,
+            _LOGGER,
+            name=f"{DOMAIN}_coordinator",
+            update_method=self.async_update_data,
```
```suggestion
```

joostlek commented on homeassistant/components/dvsportal/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+        )
+        self.dvs_portal = dvs_portal
+
+    async def async_update_data(self) -> DVSPortalData:
```
```suggestion
    async def _async_update_data(self) -> DVSPortalData:
```

joostlek commented on homeassistant/components/dvsportal/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+        """Fetch data from the DVSPortal API."""
+        try:
+            await self.dvs_portal.update()
+        except Exception as e:
```
You're only allowed to catch exceptions in the config flow

ChessSpider committed:
Update homeassistant/components/dvsportal/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

ChessSpider committed:
Update homeassistant/components/dvsportal/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

ChessSpider committed:
Update homeassistant/components/dvsportal/coordinator.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

ChessSpider committed:
Update homeassistant/components/dvsportal/coordinator.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

ChessSpider committed:
resolve comments from Joost

ChessSpider committed:
more dataclasses

ChessSpider committed:
forgot this one..

ChessSpider committed:
do not actch exceptins in coordinator update()

ChessSpider commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,108 @@
+
+    dvs_portal: DVSPortal
+    coordinator: DVSPortalCoordinator
+    ha_registered_license_plates: set[str]
```
I removed dvs_porta.

ha_registered_license_plates is used in sensor.py to store all the license places which `DVSCarSensor` has been created; to prevent they get created multiple times.

ChessSpider commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,108 @@
+        await coordinator.async_config_entry_first_refresh()
+    except dvs_exceptions.DVSPortalError as ex:
+        await dvs_portal.close()
+        raise exceptions.ConfigEntryNotReady("Failed to initialize DVSPortal") from ex
```
i restructured the try/catch

ChessSpider committed:
add comment for future me

ChessSpider commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,108 @@
+        await dvs_portal.close()
+        raise exceptions.ConfigEntryNotReady("Failed to initialize DVSPortal") from ex
+
+    async def async_unload_entry(
```
restructured

ChessSpider commented:
Hi @joostlek ,

I believe I've addressed all your feedback on this MR. If there's anything else that needs improvement, I should have time to work on it this weekend.

Also, I'd like to work on getting the service calls back in the MR, as that's where the true value of this integration lies. Could you share your thoughts on how we can approach that?

Let me know what you think!

ChessSpider committed:
Merge branch 'dev' into dvsportal

frenck committed:
Merge branch 'dev' into dvsportal

joostlek commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,106 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
```
```suggestion
```

joostlek commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,106 @@
+    """Data class for runtime data of DVSPortal."""
+
+    coordinator: DVSPortalCoordinator
+    ha_registered_license_plates: set[
```
why the `ha_`?

joostlek commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,106 @@
+async def async_setup_entry(hass: HomeAssistant, entry: DVSPortalConfigEntry) -> bool:
+    """Set up the dvsportal component from a config entry."""
+
+    async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        """Unload a config entry."""
+        runtime_data: DVSPortalRuntimeData = entry.runtime_data
+
+        # Ensure dvs_portal.close() is called to clean up the session
+        if dvs_portal := runtime_data.coordinator.dvs_portal:
+            try:
+                await dvs_portal.close()
+            except Exception as ex:  # noqa: BLE001
+                _LOGGER.warning("Failed to close DVSPortal session: %s", ex)
+
+        unload_ok = await hass.config_entries.async_forward_entry_unload(
+            entry, "sensor"
+        )
+        if unload_ok:
+            hass.data[DOMAIN].pop(entry.entry_id)
+
+        return unload_ok
```
why is this inside of `async_setup_entry`?

joostlek commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,106 @@
+
+    except dvs_exceptions.DVSPortalError as ex:
+        await dvs_portal.close()
+        raise exceptions.ConfigEntryNotReady("Failed to initialize DVSPortal") from ex
```
```suggestion
        raise ConfigEntryNotReady("Failed to initialize DVSPortal") from ex
```

joostlek commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,106 @@
+        coordinator=coordinator,
+        ha_registered_license_plates=set(),
+    )
+    await hass.config_entries.async_forward_entry_setups(entry, ["sensor"])
```
Please use the `Platform` enum

joostlek commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,106 @@
+    )
+    await hass.config_entries.async_forward_entry_setups(entry, ["sensor"])
+
+    # Check if the first refresh is successful
+    await coordinator.async_config_entry_first_refresh()
```
do the refresh before forwarding the setups

joostlek commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,106 @@
+        )
+    )
+
+    entry.add_update_listener(async_update_options)
```
we don't need this update listener as we don't have anything to update

joostlek commented on homeassistant/components/dvsportal/__init__.py:
```diff
@@ -0,0 +1,106 @@
+    # Check if the first refresh is successful
+    await coordinator.async_config_entry_first_refresh()
+
+    entry.async_on_unload(
+        async_dispatcher_connect(
+            hass, f"{DOMAIN}_{entry.entry_id}_unload", async_unload_entry
+        )
+    )
```
why?

joostlek commented on homeassistant/components/dvsportal/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+    """Error to indicate there is invalid auth."""
+
+
+class DVSPortalConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
```suggestion
class DVSPortalConfigFlow(ConfigFlow, domain=DOMAIN):
```

joostlek commented on homeassistant/components/dvsportal/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+
+    VERSION = 1
+
+    async def async_step_user(self, user_input=None) -> config_entries.ConfigFlowResult:
```
```suggestion
    async def async_step_user(self, user_input: dict[str, Any] | None = None) -> ConfigFlowResult:
```

joostlek commented on homeassistant/components/dvsportal/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+        if user_input is not None:
+            try:
+                info = await self.validate_input(user_input)
+                return self.async_create_entry(title=info["title"], data=user_input)
```
Only have things in the try block that can raise

joostlek commented on homeassistant/components/dvsportal/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+            except AbortFlow:
+                raise
+            except Exception:  # pylint: disable=broad-except
+                logging.exception("Error in async step user")
```
```suggestion
                _LOGGER.exception("Unknown error")
```

joostlek commented on homeassistant/components/dvsportal/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+                errors["base"] = "invalid_auth"
+            except AbortFlow:
+                raise
+            except Exception:  # pylint: disable=broad-except
```
```suggestion
            except Exception:
```

joostlek commented on homeassistant/components/dvsportal/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+        except DVSPortalAuthError:
+            raise InvalidAuth from None
+
+        return {"title": identifier}
```
we don't have to return anything tbh

Since this function is only used once, I would rather inline it in the user step

joostlek commented on homeassistant/components/dvsportal/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+        password = data[CONF_PASSWORD]
+        user_agent = data.get("user_agent", "HomeAssistant")
+
+        await self.async_set_unique_id(f"{api_host}.{identifier}")
```
the api host generally isnt unique, why is it now?

joostlek commented on homeassistant/components/dvsportal/entity.py:
```diff
@@ -0,0 +1,103 @@
+        )
+
+    def _set_state(self):
+        """Set the current state of this car (license plate).
+
+        The state can be present, reserved or not present.
+        """
+        reservation = self.coordinator.data.active_reservations.get(self._license_plate)
+        if reservation is None:
+            self._reset_attributes()
+            self._attr_state = "not present"
+        else:
+            self._attr_extra_state_attributes.update(reservation)
+
+            now = datetime.now()
+            valid_until = datetime.strptime(
+                str(reservation.get("valid_until", "1900-01-01T00:00:00")).split(
+                    ".", maxsplit=1
+                )[0],
+                "%Y-%m-%dT%H:%M:%S",
+            )
+            valid_from = datetime.strptime(
+                str(reservation.get("valid_from", "1900-01-01T00:00:00")).split(
+                    ".", maxsplit=1
+                )[0],
+                "%Y-%m-%dT%H:%M:%S",
+            )
+
+            if valid_from <= now < valid_until:
+                self._attr_state = "present"
+            else:
+                self._attr_state = "reserved"
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._reset_attributes()
+        self._set_state()
+
+        self.async_write_ha_state()
```
I think I am missing what this class is doing. base entities should be a way to bootstrap your entities with logic shared between platforms, like update mechanisms, state handling or static attributes, and this is doing a lot more that should not be here

ChessSpider commented:
hI @joostlek , thank you for the additional feedback. I have decided to keep this as a custom component for now. This is more effort than I anticipated, especially as the main service calls are not even part of the MR anymore.

Ill close the related MRs.
